src/main/java/com/minecolonies/coremod/quests/QuestManager.java

```diff
@@ -0,0 +1,220 @@
+package com.minecolonies.coremod.quests;
+
+import com.minecolonies.api.colony.IColony;
+import com.minecolonies.api.quests.IColonyQuest;
+import com.minecolonies.api.quests.IQuestData;
+import com.minecolonies.api.quests.IQuestManager;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.entity.player.Player;
+
+import java.util.*;
+
+import static com.minecolonies.api.util.constant.NbtTagConstants.*;
+
+/**
+ * Quest manager of each colony.
+ */
+public class QuestManager implements IQuestManager
+{
+    /**
+     * All quests that have been unlocked.
+     */
+    private final Map<ResourceLocation, IColonyQuest> availableQuests = new HashMap<>();
+
+    /**
+     * All quests that have been finished. And how often.
+     */
+    private final Map<ResourceLocation, Integer> finishedQuests = new HashMap<>();
+
+    /**
+     * All quests in progress.
+     */
+    private final Map<ResourceLocation, IColonyQuest> inProgressQuests = new HashMap<>();
+
+    /**
+     * This manager's colony
+     */
+    private final IColony colony;
+
+    public QuestManager(final IColony colony)
+    {
+        this.colony = colony;
+    }
+
+    @Override
+    public boolean attemptAcceptQuest(final ResourceLocation questID, final Player player)
+    {
+        final IColonyQuest quest = availableQuests.getOrDefault(questID, null);
+        if (quest == null || !quest.isValid(colony))
+        {
+            return false;
+        }
+        this.inProgressQuests.put(questID, quest);
+        this.availableQuests.remove(questID);
+        return true;
+    }
+
+    @Override
+    public void concludeQuest(final ResourceLocation questId)
+    {
+        if (inProgressQuests.containsKey(questId))
+        {
+            inProgressQuests.remove(questId);
+            finishedQuests.put(questId, finishedQuests.getOrDefault(questId, 0) + 1);
+        }
+        else if (availableQuests.containsKey(questId))
+        {
+            // When a player short-cut quits a job without accepting it. (E.g. been there, done that options).
+            availableQuests.remove(questId);
+            finishedQuests.put(questId, finishedQuests.getOrDefault(questId, 0) + 1);
```

Comment:
Should this be counted as "finished" if they skipped it?

Reply:
Yes. So that it doesnt show up anymore, etc.

Reply:
but still unlocks the fullowups.

Reply:
But why unlock the followup?  Someone has a quest "please deliver this apple for me", you respond "no, i don't have time, go away".  Then the target gives a quest "thanks for giving me that apple from X, can you give them this diamond in response?"  It makes no sense.

Reply:
Well, this is up to the quest designer to design this. Else they just make it a cancel not a skip.